### PR TITLE
295 automate submission fields

### PR DESCRIPTION
**Pull request by marcellmueller**
https://app.zenhub.com/workspaces/ccbc---delivery-board-61c0c1204c8bcb001491c5f3/issues/bcgov/conn-ccbc-portal/295

### CODE REVIEW

File: app/formSchema/pages/submission.ts [hunk 1]
```diff
@@ -26,6 +22,10 @@ const submission = {
         title: 'Title',
         type: 'string',
       },
+      submissionDate: {
+        title: 'On this date (YYYY-MM-DD)',
+        type: 'string',
+      },
```
Comment (matthieu-foucault) [inline app/formSchema/pages/submission.ts:28]: Is this change needed?

File: app/formSchema/uiSchema/pages/submission.ts [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
+    'submissionDate',
   ],
   'ui:title': '',
   'ui:field': 'SubmissionField',
+  submissionCompletedFor: {
+    'ui:widget': ReadOnlySubmissionWidget,
+  },
   submissionDate: {
-    'ui:widget': 'DatePickerWidget',
+    'ui:widget': ReadOnlySubmissionWidget,
```
Comment (matthieu-foucault) [inline app/formSchema/uiSchema/pages/submission.ts:16]:
```suggestion
    'ui:widget': 'ReadOnlySubmissionWidget',
```

Widgets should be strings and added to the theme, otherwise themes won't work

File: app/lib/theme/widgets/ReadOnlySubmissionWidget.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        {value ? (
+          value
+        ) : (
+          <StyledError>
+            No legal organization name was provided. Please return to the{' '}
+            <StyledLink href={`/form/${id}/${pageNumber}`}>
+              Organization Profile
+            </StyledLink>{' '}
+            page and enter one.
+          </StyledError>
```
Comment (matthieu-foucault) [inline app/lib/theme/widgets/ReadOnlySubmissionWidget.tsx:45]: Are we missing a test for this?

File: db/deploy/mutations/submit_application.sql [hunk 4]
```diff
@@ -24,8 +28,31 @@ begin
...
+  select form_data -> 'submission' ->> 'submissionCompletedFor',
+    form_data -> 'submission' ->> 'submissionCompletedBy',
+    form_data -> 'submission' ->> 'submissionTitle',
+    form_data -> 'submission' ->> 'submissionDate'
+   from ccbc_public.application where id = application_row_id
+   into submission_completed_for, submission_completed_by, submission_title, submission_date;
+
+  if coalesce(submission_completed_for, '') = '' then
+    raise 'The application cannot be submitted as the submission field submission_completed_for is null or empty';
+end if;
```
Comment (matthieu-foucault) [inline db/deploy/mutations/submit_application.sql:40]:
```suggestion
  end if;
```

File: db/sqitch.plan [hunk 5]
```diff
@@ -32,3 +32,4 @@ mutations/create_application 2022-08-31T16:16:23Z Anthony Bushara <anthony@butto
 @1.0.1 2022-09-02T00:40:17Z Matthieu Foucault <matthieu@button.is> # release v1.0.1
 @1.1.0 2022-09-06T23:59:22Z Matthieu Foucault <matthieu@button.is> # release v1.1.0
 @1.1.1 2022-09-09T20:35:47Z Marcel Mueller,,, <marcel@Thinkpad> # release v1.1.1
+mutations/submit_application [mutations/submit_application@1.0.0] 2022-09-13T19:52:43Z Marcel Mueller,,, <marcel@button.is> # Check all submission fields in form_data are filled before submission
```
Comment (matthieu-foucault) [inline db/sqitch.plan:35]:
```suggestion
mutations/submit_application [mutations/submit_application@1.0.0] 2022-09-13T19:52:43Z Marcel Mueller <marcel@button.is> # Check all submission fields in form_data are filled before submission
```

You should configure sqitch so that the correct user name is added in the plan

File: db/test/unit/mutations/submit_application_test.sql [hunk 6]
```diff
@@ -89,6 +98,34 @@ select results_eq(
   'Increases the ccbc number when submitting applications'
 );
 
+select throws_like(
+  $$
+    select ccbc_public.submit_application(5)
+  $$,
+  'The application cannot be submitted as the submission field submission_date is null or empty'
```
Comment (matthieu-foucault) [inline db/test/unit/mutations/submit_application_test.sql:105]:
```suggestion
  'The application cannot be submitted as the submission field submission_date is null or empty',
  'Throws an exception if submission_date is empty'
```
Adding test descriptions will improve logging if the tests fail

File: db/revert/mutations/submit_application@1.0.0.sql [hunk 7]
```diff
@@ -0,0 +1,7 @@
+-- Revert ccbc:mutations/submit_application from pg
```
Comment (matthieu-foucault) [inline db/revert/mutations/submit_application@1.0.0.sql:1]: Where is the revert created by `sqitch rework`? There should be a revert that resets the function to its v1.0.0 behaviour

**Review by matthieu-foucault — CHANGES_REQUESTED** [review]

File: app/formSchema/pages/submission.ts (hunk 1 above)
Comment (marcellmueller) [reply app/formSchema/pages/submission.ts:28]: I prefer to keep it in the same order as it appears on the form itself or the `ui:schema`, it's more readable that way.

File: db/test/unit/mutations/submit_application_test.sql (hunk 6 above)
Comment (marcellmueller) [reply db/test/unit/mutations/submit_application_test.sql:105]: I don't see how this is any more helpful or descriptive than what is already there, which already shows in the logs when a test fails.

Comment (matthieu-foucault) [reply db/test/unit/mutations/submit_application_test.sql:105]: Fair, I assumed the log was more cryptic without a drescription

File: app/components/Layout.tsx [hunk 8]
```diff
@@ -17,7 +17,7 @@ const StyledLayout = styled('div')`
   min-height: 100vh;
 `;
 const StyledMain = styled('main')`
-  margin: auto 50px;
+  margin: 50px auto;
```
Comment (matthieu-foucault) [inline app/components/Layout.tsx:20]: Why is this changed? This is making pages such as organization location very narrow. If there are overall layout fixes to make, they belong in a separate PR

**Review by matthieu-foucault — CHANGES_REQUESTED** [review]

File: app/components/Layout.tsx (hunk 8 above)
Comment (marcellmueller) [reply app/components/Layout.tsx:20]: oh that was a mistake, I was playing with something because I noticed the layout isn't centered anymore. I guess our happo setup doesn't catch things like that. I'll remove it though.

Comment (matthieu-foucault) [reply app/components/Layout.tsx]: Happo did catch it, but we have so many spurious diffs that we're blindly approving happo currently

File: app/lib/theme/widgets/ReadOnlySubmissionWidget.tsx [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+      <StyledValue>
+        {value}
+        {submissionCompletedFor && !value && (
+          <StyledError>
+            No legal organization name was provided. Please return to the{' '}
+            <StyledLink href={`/form/${id}/${pageNumber}`}>
+              Organization Profile
+            </StyledLink>{' '}
+            page and enter one.
+          </StyledError>
```
Comment (matthieu-foucault) [inline app/lib/theme/widgets/ReadOnlySubmissionWidget.tsx:51]: This pattern of having field-specific branching behaviour within widgets is at odds with having a declarative uiSchema. If we need specific behaviour for a single field, we should either create a component for it, or find a generic way to have this custom behaviour. 

Can the [custom validation](https://react-jsonschema-form.readthedocs.io/en/docs/usage/validation/#custom-validation-rules) add some JSX as an error?
A generic solution could be a `customError` uiSchema option that would be handled by our field template.

If we do need a custom component, in this case it should be a custom field, not a widget. Widgets should only concern themselves with rendering an input/value.

File: app/tests/components/Form/ApplicationForm.test.tsx [hunk 10]
```diff
@@ -26,6 +26,29 @@ const mockQueryPayload = {
...
+  Query() {
+    return {
+      openIntake: {
+        closeTimestamp: '2022-08-27T12:51:26.69172-04:00',
+      },
+    };
+  },
+};
+
+const submissionPayload = {
```
Comment (matthieu-foucault) [inline app/tests/components/Form/ApplicationForm.test.tsx:41]:
```suggestion
const submissionPayload = {
  ...mockQueryPayload,
```

This way you don't need to copy the `Query()` function every time

File: app/tests/components/Form/ApplicationForm.test.tsx [hunk 11]
```diff
@@ -166,41 +254,25 @@ describe('The application form', () => {
   });
 
   it('submission page submit button is enabled on when all inputs filled', async () => {
-    componentTestingHelper.loadQuery();
+    componentTestingHelper.loadQuery(submissionPayload);
```
Comment (matthieu-foucault) [inline app/tests/components/Form/ApplicationForm.test.tsx:257]: `submissionPayload` doesn't have the `submissionCompletedFor` and `submissionDate`, should they be present for the button to be enabled?

File: app/tests/components/Form/ApplicationForm.test.tsx [hunk 12]
```diff
@@ -166,41 +254,25 @@ describe('The application form', () => {
...
-          screen
-            .getByRole('button', { name: 'Submit' })
-            .hasAttribute('disabled')
-        ).toBeFalse()
-      );
+      expect(
+        screen.getByRole('button', { name: 'Submit' }).hasAttribute('disabled')
+      ).toBeFalse();
     });
   });
```
Comment (matthieu-foucault) [inline app/tests/components/Form/ApplicationForm.test.tsx:278]: Looks like the test below has the wrong title?

**Review by matthieu-foucault — CHANGES_REQUESTED** [review]

File: app/tests/components/Form/ApplicationForm.test.tsx (hunk 10 above)
Comment (marcellmueller) [reply app/tests/components/Form/ApplicationForm.test.tsx:41]: Man you're being way too opinionated. This is getting annoying.

Comment (matthieu-foucault) [reply app/tests/components/Form/ApplicationForm.test.tsx:41]: In what way proposing solutions following best practices to reduce code duplication is being opinionated?

Comment (marcellmueller) [reply app/tests/components/Form/ApplicationForm.test.tsx:41]: I mean most of this review in general.

Comment (matthieu-foucault) [reply app/tests/components/Form/ApplicationForm.test.tsx:41]: Feel free to discuss any feedback on the relevant threads. I'm happy to expand on any of my suggestions. Some of them are probably beyond the scope of this PR and should be tracked in a separate issue

File: app/formSchema/uiSchema/pages/submission.ts [hunk 13]
```diff
@@ -1,14 +1,20 @@
...
     'submissionCompletedBy',
     'submissionTitle',
+    'submissionDate',
   ],
   'ui:title': '',
   'ui:field': 'SubmissionField',
+  submissionCompletedFor: {
+    'ui:widget': 'ReadOnlySubmissionWidget',
+    'ui:options': {
+      'field-name': 'submissionCompletedFor',
```
[the remark this answers is not in the document]
Comment (marcellmueller) [reply app/formSchema/uiSchema/pages/submission.ts:13]: I made a custom read only widget for submission page because the design has different styling than the other read only widget (no padding, not bold). Rather than make a ton of `ui:options` flags it made sense to make one just for this page as it is the only place in the design that has it.

[the remark this answers is not in the document]
Comment (matthieu-foucault) [reply app/formSchema/uiSchema/pages/submission.ts:13]: If there's a different behaviour just for this page, then it makes sense to have a custom widget instead of using `readonly` (I'd like to know why our designs aren't consistent across pages though). 
I suppose that rendering the error within the widget makes sense in this case because we're not displaying other errors, just disabling the submit field (which is a separate rabbit hole we won't go into here...).

I still think having behaviour/ui options in the rjsf fields/widgets that is specific to one field is an antipattern and should be avoided. It's unclear when looking at the ui schema that this option will make the `ReadOnlySubmissionWidget` change its behaviour to something that is custom for this specific field. If one field needs a component to display something for that field only, let's make a widget for that field only

File: app/lib/theme/widgets/ReadOnlySubmissionWidget.tsx [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+const ReadOnlySubmissionWidget: React.FC<WidgetProps> = ({
+  options,
+  value,
+}) => {
+  const router = useRouter();
+  const { id } = router.query;
+
+  const pageNumber =
+    Object.keys(a.properties).indexOf('organizationProfile') + 1;
```
[the remark this answers is not in the document]
Comment (marcellmueller) [reply app/lib/theme/widgets/ReadOnlySubmissionWidget.tsx:35]: We should probably have a utility function for this but I am doing it the same way we have been doing it for ages in `ApplicationForm.tsx` here.

File: app/cypress/integration/index.spec.js [hunk 15]
```diff
@@ -468,11 +468,6 @@ context('Homepage', () => {
 
     cy.get('h1').contains('Submission');
 
-    cy.get('input[id="root_submissionCompletedFor"]').type('test');
-
-    cy.get('#root_submissionDate').click();
-    cy.get(`.react-datepicker__day--001`).eq(2).click({ force: true });
-
```
[the remark this answers is not in the document]
Comment (marcellmueller) [reply app/cypress/integration/index.spec.js:475]: You previously told me that this end to end test was supposed to follow the happy path and didn't need to test every input?

[the remark this answers is not in the document]
Comment (matthieu-foucault) [reply app/cypress/integration/index.spec.js:475]: I was debating commenting on this one, but I think there's a difference between reducing our test coverage vs adding perfect tests: while changes don't need to have perfect coverage, they shouldn't reduce existing tests coverage

In particular here, since the feature relies on something that happens on another page (inserting the `submissionCompletedFor` and `submissionDate` when the other forms are submitted), I think this is something that would benefit from an end-to-end test.

[the remark this answers is not in the document]
Comment (marcellmueller) [reply app/cypress/integration/index.spec.js:475]: Okay I’ll add a check for the value tomorrow.

File: db/sqitch.plan [hunk 16]
```diff
@@ -31,5 +31,6 @@ mutations/create_application 2022-08-31T16:16:23Z Anthony Bushara <anthony@butto
 @1.0.0 2022-09-01T23:42:33Z Matthieu Foucault <matthieu@button.is> # release v1.0.0
 @1.0.1 2022-09-02T00:40:17Z Matthieu Foucault <matthieu@button.is> # release v1.0.1
 @1.1.0 2022-09-06T23:59:22Z Matthieu Foucault <matthieu@button.is> # release v1.1.0
-@1.1.1 2022-09-09T20:35:47Z Marcel Mueller,,, <marcel@Thinkpad> # release v1.1.1
+@1.1.1 2022-09-09T20:35:47Z Marcel Mueller <marcel@button.is> # release v1.1.1
```
Comment (matthieu-foucault) [inline db/sqitch.plan:34]: This line has been deployed to prod and cannot be updated, otherwise the change will have a different id, leading to the deploy failing.

File: app/tests/components/Form/ReadOnlySubmissionWidget.test.tsx [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+const uiSchema = {
+  submissionDate: {
+    'ui:widget': 'ReadOnlySubmissionWidget',
+  },
+};
+
+const renderStaticLayout = (
+  schema: JSONSchema7,
+  uiSchema: JSONSchema7,
```
Comment (matthieu-foucault) [inline app/tests/components/Form/ReadOnlySubmissionWidget.test.tsx:25]:
```suggestion
  uiSchema: any,
```
That, or the uiSchema type exported by rjsf, if there is one. uiSchema isn't a json schema

File: app/tests/components/Form/ReadOnlySubmissionWidget.test.tsx [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+      onSubmit={() => console.log('test')}
+      schema={schema as JSONSchema7}
+      uiSchema={uiSchema}
+    />
+  );
+};
+
+describe('The ReadOnlySubmissionWidget', () => {
+  it('should render the title', () => {
+    renderStaticLayout(schema as JSONSchema7, uiSchema as JSONSchema7, {});
```
Comment (matthieu-foucault) [inline app/tests/components/Form/ReadOnlySubmissionWidget.test.tsx:40]: Instead of a cast here, typing the `schema` variable above would give you better typing. The cast will effectively silence the type checking.

**Review by matthieu-foucault — CHANGES_REQUESTED** [review]
The sqitch change is required, otherwise the deploy will fail.
The test file change are typescript details and quick fixes.

**Review by matthieu-foucault — APPROVED** [review]
Great work. Thanks for making the requested changes
